core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1327,69 +1327,337 @@ class ReassignedPartitionsIsrChangeListener(protected val controller: KafkaContr
+    }
+
+    private def getTopicAndPartition(child: String): Set[TopicAndPartition] = {
+      val changeZnode: String = ZkUtils.IsrChangeNotificationPath + "/" + child
+      val (jsonOpt, _) = controllerContext.zkUtils.readDataMaybeNull(changeZnode)
+      if (jsonOpt.isDefined) {
+        val json = Json.parseFull(jsonOpt.get)
+
+        json match {
+          case Some(m) =>
+            val topicAndPartitions: mutable.Set[TopicAndPartition] = new mutable.HashSet[TopicAndPartition]()
+            val isrChanges = m.asInstanceOf[Map[String, Any]]
+            val topicAndPartitionList = isrChanges("partitions").asInstanceOf[List[Any]]
+            topicAndPartitionList.foreach {
+              case tp =>
+                val topicAndPartition = tp.asInstanceOf[Map[String, Any]]
+                val topic = topicAndPartition("topic").asInstanceOf[String]
+                val partition = topicAndPartition("partition").asInstanceOf[Int]
+                topicAndPartitions += TopicAndPartition(topic, partition)
+            }
+            topicAndPartitions
+          case None =>
+            error("Invalid topic and partition JSON: " + jsonOpt.get + " in ZK: " + changeZnode)
+            Set.empty
+        }
+      } else {
+        Set.empty
+      }
+    }
+
+  }
+
+  case class PreferredReplicaLeaderElection(partitionsForPreferredReplicaElection: Set[TopicAndPartition]) extends ControllerEvent {
+    override def process(): Unit = {
+      if (!isActive) return
+      if (controllerContext.partitionsUndergoingPreferredReplicaElection.nonEmpty)
+        info("These partitions are already undergoing preferred replica election: %s"
+          .format(controllerContext.partitionsUndergoingPreferredReplicaElection.mkString(",")))
+      val partitions = partitionsForPreferredReplicaElection -- controllerContext.partitionsUndergoingPreferredReplicaElection
+      val partitionsForTopicsToBeDeleted = partitions.filter(p => topicDeletionManager.isTopicQueuedUpForDeletion(p.topic))
+      if (partitionsForTopicsToBeDeleted.nonEmpty) {
+        error("Skipping preferred replica election for partitions %s since the respective topics are being deleted"
+          .format(partitionsForTopicsToBeDeleted))
       }
+      onPreferredReplicaElection(partitions -- partitionsForTopicsToBeDeleted)
     }
   }
 
-  private def processUpdateNotifications(topicAndPartitions: immutable.Set[TopicAndPartition]) {
-    val liveBrokers: Seq[Int] = controller.controllerContext.liveOrShuttingDownBrokerIds.toSeq
-    debug("Sending MetadataRequest to Brokers:" + liveBrokers + " for TopicAndPartitions:" + topicAndPartitions)
-    controller.sendUpdateMetadataRequest(liveBrokers, topicAndPartitions)
+  case object AutoPreferredReplicaLeaderElection extends ControllerEvent {
+    override def process(): Unit = {
+      scheduleAutoLeaderRebalanceTask(delay = config.leaderImbalanceCheckIntervalSeconds, unit = TimeUnit.SECONDS)
```

Comment:
Since we aren't letting the scheduler periodically inject the event into the queue, any single manual event injection we miss means that there will be no later injection until we become re-elected as controller. This is another reason why I wanted the scheduler to do the periodic event injections.

We risk skipping the next `AutoPreferredReplicaLeaderElection` if we naively put it after `checkAndTriggerPartitionRebalance()`, as `checkAndTriggerPartitionRebalance()` could throw an exception.

Putting the schedule line before the `checkAndTriggerPartitionRebalance()` prevents this from happening.

We can alternatively wrap the call to `checkAndTriggerPartitionRebalance()` in a try/finally and put the schedule call in the finally block as well.

I'm open to doing the try/finally, keeping the code as is, or reverting the logic to just let the scheduler periodically inject the event.